feat(ListInline): add List and ListInlineItem components

johnathanmiller opened the pull request:
<!---
What types of changes does your code introduce? Put an `x` in all the boxes that apply:
-->

- [ ] Bug fix <!-- (change which fixes an issue) -->
- [x] New feature <!-- (change which adds functionality) -->
- [ ] Chore <!-- (change which doesn't affect the usage of the package (such as documentation, build process, or project setup changes)) -->
- [ ] Breaking change <!-- (fix or feature that would cause existing functionality to change) -->
- [x] There is an open issue which this change addresses
- [x] I have read the **[CONTRIBUTING](./CONTRIBUTING.md)** document.
- [x] My commits follow the [Git Commit Guidelines](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#commits)
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly.
- [x] My change requires a change to [Typescript typings](./types/lib).
  - [x] I have updated the typings accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

#2031

kyletsang commented on src/ListInlineItem.js:
```diff
@@ -0,0 +1,36 @@
+    tag: 'li'
+};
+
+const ListInlineItem = (props) => {
```
Let's also use a `forwardRef` here

kyletsang commented on src/ListInline.js:
```diff
@@ -0,0 +1,36 @@
+    tag: 'ul'
+};
+
+const ListInline = (props) => {
```
Let's make this component a little bit more generic so we can cover `.list-unstyled` here too, for example:
`<List type="unstyled">` and `<List type="inline">`

Also use a `forwardRef` here

johnathanmiller commented:
Summary of changes:
- Renamed `ListInline` to `List`.
- `List` now accepts a prop of `type`
- `List` and `ListInlineItem` both accept `ref`
- Docs now have a `List` page which contains both `unstyled` and `inline` list examples
- Updated tests

kyletsang commented on src/List.js:
```diff
@@ -0,0 +1,40 @@
+  className: PropTypes.string,
+  cssModule: PropTypes.object,
+  type: PropTypes.string,
+  innerRef: PropTypes.oneOfType([PropTypes.object, PropTypes.string, PropTypes.func])
```
Let's use `forwardRef` instead of `innerRef` since this is more suited for function components:
https://reactjs.org/docs/forwarding-refs.html

So below would be
`const List= React.forwardRef((props, ref) => ...`

kyletsang commented:
Looks good.  One final fix for the `forwardRef` for both components and we should be good to go.

kyletsang reviewed (approved):
Looks good, thanks!
